### PR TITLE
CUDA-aware MPI

### PR DESCRIPTION
**Pull request by masterleinad**
I still need to clean this up. In particular, the preprocessor variables are not going to stay, but will be replaced by a `CMake` option. So far, I was testing this with
- 60 MPI processes, 
- 10^7 points/MPI process,
- 10^6 queries/MPI process,
- and a varying number of neighbors for the knn search,
- and CUDA

using the `distributed_tree_driver` benchmark.

|  neighbors | old   |  new   |
|------------|-------|--------|
|  10        | 2.4e0 |  3.7e0 |
| 20         | 4.9e0 | 5.8e0  |
| 40         | 1.0e1 | 1.0e1  |
| 80         | 2.2e1 | 2.0e1  |
| 100        | 2.8e1 | 2.5e1  |
| 120        | 3.3e1 | 3.0e1  |
| 140        | 4.0e1 | 3.4e1  |
| 160        | 4.6e1 | 3.9e1  |

### CODE REVIEW

**Comment by dalg24** [comment]
Both CUDA builds were hanging so I ended up killing the build.  @masterleinad please investigate locally and make sure we don't generate traffic jams on the Jenkins server.

**Comment by masterleinad** [comment]
That hopefully works now. I was missing a `cudaDeviceSynchronize` before calling the MPI functions.

File: src/ArborX_DistributedSearchTree.hpp [hunk 1]
```diff
@@ -121,8 +121,9 @@ DistributedSearchTree<DeviceType>::DistributedSearchTree(
   Kokkos::View<Box *, DeviceType> boxes(
       Kokkos::ViewAllocateWithoutInitializing("rank_bounding_boxes"),
       comm_size);
-  // FIXME when we move to MPI with CUDA-aware support, we will not need to
-  // copy from the device to the host
+  // We could avoid copying between device host and device with CUDA-aware MPI,
+  // but this requires the object to be initialized on the device with host data
+  // (_bottom_tree.bounds()). Currently, it doesn't seem worthwhile trying this.
```
Comment (dalg24) [inline src/ArborX_DistributedSearchTree.hpp]: Try with device data.  Here is how you can get it.
```C++
using TreeAccess = Details::TreeVisualization<DeviceType>::TreeAccess;
TreeAccess::getBoundingVolume(TreeAccess::getRoot(_bottom_tree), _bottom_tree);
```
PS: I am not advocating that is good design but at least it will tell us if this is worth pursuing at all.

**Comment by dalg24** [comment]
image

File: src/ArborX_DistributedSearchTree.hpp (hunk 1 above)
Comment (dalg24) [reply src/ArborX_DistributedSearchTree.hpp]: This will be explored in a separate PR.

Comment (masterleinad) [reply src/ArborX_DistributedSearchTree.hpp]: See #174.

File: src/details/ArborX_DetailsDistributor.hpp [hunk 2]
```diff
@@ -173,6 +186,12 @@ class Distributor
                   _comm, &requests.back());
       }
     }
+
+#ifdef ARBORX_USE_CUDA_AWARE_MPI
+    if (std::is_same<ExecutionSpace, Kokkos::Cuda>::value)
+      cudaDeviceSynchronize();
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributor.hpp]: Why this  rather than unconditional fence on the execution space?

File: benchmarks/distributed_tree_driver/distributed_tree_driver.cpp [hunk 3]
```diff
@@ -477,7 +477,7 @@ int main(int argc, char *argv[])
     else if (node == "cuda")
     {
 #ifdef KOKKOS_ENABLE_CUDA
-      using Node = Kokkos::Cuda;
+      using Node = Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>;
```
Comment (dalg24) [inline benchmarks/distributed_tree_driver/distributed_tree_driver.cpp]: This change does not belong to this PR.  At the very least should be its own commit.

Comment (masterleinad) [reply benchmarks/distributed_tree_driver/distributed_tree_driver.cpp]: It has its own commit now.

**Comment by masterleinad** [comment]
`CUDA-aware MPI` is off by default. Do we want to test for it in one of the configurations?
